### PR TITLE
Add custom stacktrace renderer which is length limit aware

### PR DESCRIPTION
**Pull request by jack-berg**
Replace `Throwable#printStackTrace(PrintStream)` exception stacktrace rendering with a custom implementation which is aware of attribute length limits, and exits early to avoid unnecessary work. The result is significantly improved memory and cpu, with gains proportional to the difference between the total stack trace length and the attribute length limit. I.e. if you have a really long stack trace (i.e. 100k+ chars) and relatively small length limit (i.e. 1k chars), you'll have more to gain with this improvement.

Benchmark results:

```
Benchmark                                            (exceptionParam)  (lengthLimit)  (renderer)  Mode  Cnt      Score      Error   Units
StacktraceRenderBenchmark.render                               SIMPLE             10         JDK  avgt    5   3341.372 ± 2897.392   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE             10         JDK  avgt    5   5650.153 ± 4700.598  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE             10         JDK  avgt    5  19064.017 ±    0.016    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE             10         JDK  avgt    5      3.000             counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE             10         JDK  avgt    5     22.000                 ms
StacktraceRenderBenchmark.render                               SIMPLE             10      CUSTOM  avgt    5     52.803 ±   30.464   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE             10      CUSTOM  avgt    5   4349.904 ± 2077.709  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE             10      CUSTOM  avgt    5    240.000 ±    0.001    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE             10      CUSTOM  avgt    5      2.000             counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE             10      CUSTOM  avgt    5      3.000                 ms
StacktraceRenderBenchmark.render                               SIMPLE           1000         JDK  avgt    5   2789.526 ± 1620.749   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE           1000         JDK  avgt    5   6961.280 ± 3623.628  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE           1000         JDK  avgt    5  20064.015 ±    0.008    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE           1000         JDK  avgt    5      3.000             counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE           1000         JDK  avgt    5      4.000                 ms
StacktraceRenderBenchmark.render                               SIMPLE           1000      CUSTOM  avgt    5    915.035 ±  482.433   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE           1000      CUSTOM  avgt    5  11918.989 ± 5471.749  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE           1000      CUSTOM  avgt    5  11312.005 ±    0.002    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE           1000      CUSTOM  avgt    5      5.000             counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE           1000      CUSTOM  avgt    5      7.000                 ms
StacktraceRenderBenchmark.render                               SIMPLE         100000         JDK  avgt    5   2483.485 ±  417.582   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE         100000         JDK  avgt    5   7305.934 ± 1181.639  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE         100000         JDK  avgt    5  19024.013 ±    0.004    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE         100000         JDK  avgt    5      3.000             counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE         100000         JDK  avgt    5      5.000                 ms
StacktraceRenderBenchmark.render                               SIMPLE         100000      CUSTOM  avgt    5   1585.589 ±  958.212   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE         100000      CUSTOM  avgt    5  11593.570 ± 5794.885  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE         100000      CUSTOM  avgt    5  18992.008 ±    0.004    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE         100000      CUSTOM  avgt    5      5.000             counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE         100000      CUSTOM  avgt    5      8.000                 ms
StacktraceRenderBenchmark.render                              COMPLEX             10         JDK  avgt    5   4291.839 ± 1596.569   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX             10         JDK  avgt    5   5042.169 ± 1731.599  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX             10         JDK  avgt    5  22568.023 ±    0.008    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX             10         JDK  avgt    5      4.000             counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX             10         JDK  avgt    5      6.000                 ms
StacktraceRenderBenchmark.render                              COMPLEX             10      CUSTOM  avgt    5     51.515 ±   24.906   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX             10      CUSTOM  avgt    5   4487.824 ± 1996.520  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX             10      CUSTOM  avgt    5    240.000 ±    0.001    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX             10      CUSTOM  avgt    5      2.000             counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX             10      CUSTOM  avgt    5      3.000                 ms
StacktraceRenderBenchmark.render                              COMPLEX           1000         JDK  avgt    5   3942.046 ±  556.053   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX           1000         JDK  avgt    5   5703.369 ±  793.320  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX           1000         JDK  avgt    5  23592.021 ±    0.004    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX           1000         JDK  avgt    5      4.000             counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX           1000         JDK  avgt    5      7.000                 ms
StacktraceRenderBenchmark.render                              COMPLEX           1000      CUSTOM  avgt    5    928.634 ±   79.862   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX           1000      CUSTOM  avgt    5  11624.281 ± 1010.560  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX           1000      CUSTOM  avgt    5  11336.005 ±    0.001    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX           1000      CUSTOM  avgt    5      6.000             counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX           1000      CUSTOM  avgt    5     10.000                 ms
StacktraceRenderBenchmark.render                              COMPLEX         100000         JDK  avgt    5   4793.245 ± 2833.184   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX         100000         JDK  avgt    5   4556.893 ± 2409.667  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX         100000         JDK  avgt    5  22552.025 ±    0.010    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX         100000         JDK  avgt    5      3.000             counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX         100000         JDK  avgt    5     11.000                 ms
StacktraceRenderBenchmark.render                              COMPLEX         100000      CUSTOM  avgt    5   2532.993 ±  638.702   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX         100000      CUSTOM  avgt    5   8701.358 ± 2035.307  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX         100000      CUSTOM  avgt    5  23088.013 ±    0.003    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX         100000      CUSTOM  avgt    5      4.000             counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX         100000      CUSTOM  avgt    5      6.000                 ms
```

### CODE REVIEW

File: sdk/common/src/test/java/io/opentelemetry/sdk/internal/StackTraceRendererTest.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                        new Exception(
+                            "cause3", new Exception("cause4", new Exception("cause5"))))))),
+        // with cause with circular reference
+        Arguments.of(withCycle),
+        // with suppressed
+        Arguments.of(withSuppressed),
+        Arguments.of(withMultipleSuppressed),
+        Arguments.of(withNestedSuppressed),
+        // with cause, cycle, and suppressed!
+        Arguments.of(withKitchenSink));
```
Comment (jack-berg) [inline sdk/common/src/test/java/io/opentelemetry/sdk/internal/StackTraceRendererTest.java:76]: Anyone else have ideas of additional exceptions to further stress the edge cases?

File: sdk/common/src/main/java/io/opentelemetry/sdk/internal/AttributeUtil.java [hunk 2]
```diff
@@ -107,7 +105,9 @@ public static Object applyAttributeLengthLimit(Object value, int lengthLimit) {
   }
 
   public static void addExceptionAttributes(
```
Comment (jack-berg) [inline sdk/common/src/main/java/io/opentelemetry/sdk/internal/AttributeUtil.java:107]: One thing I've considered is retaining some sort of system property / env var to fallback to the built-in JDK exception rendering. Would offer a nice escape hatch to revert to the built-in jdk rendering in case there's any bugs in the code, but downside is it would be hard to know when it would be safe to finally remove.

File: sdk/common/src/jmh/java/io/opentelemetry/sdk/internal/StacktraceRenderBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    @Param({"10", "1000", "100000"})
+    int lengthLimit;
+  }
+
+  @SuppressWarnings("ImmutableEnumChecker")
+  public enum Renderer {
+    JDK(
+        (throwable, limit) -> {
+          StringWriter stringWriter = new StringWriter();
+          try (PrintWriter printWriter = new PrintWriter(stringWriter)) {
```
Comment (fandreuz) [inline sdk/common/src/jmh/java/io/opentelemetry/sdk/internal/StacktraceRenderBenchmark.java:62]: Not sure if the overhead of opening a `PrintWriter` is negligible here, but you may want to move initialization and cleanup into two `@Setup` and `@TearDown` methods

File: sdk/common/src/main/java/io/opentelemetry/sdk/internal/StackTraceRenderer.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  private void appendStackTrace() {
+    if (appendLine(throwable.toString())) {
+      return;
+    }
+
+    StackTraceElement[] stackTraceElements = throwable.getStackTrace();
+    for (StackTraceElement stackTraceElement : stackTraceElements) {
+      if (appendLine("\tat " + stackTraceElement)) {
```
Comment (fandreuz) [inline sdk/common/src/main/java/io/opentelemetry/sdk/internal/StackTraceRenderer.java:49]: You may be able to squeeze a bit more ops/sec by replacing string concatenations all around this class with `StringBuilder#append`. I've tried to run the benchmark, and it seems that the [patch](https://github.com/user-attachments/files/19802914/patch.txt) has some impact (e.g. `4960 ns/op` vs `3944 ns/op` for the bulkiest test).

### Before
```
StacktraceRenderBenchmark.render                               SIMPLE             10         JDK  avgt    5   3085.376 ±  3127.418   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE             10         JDK  avgt    5   6087.686 ±  5313.686  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE             10         JDK  avgt    5  18824.056 ±     1.163    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE             10         JDK  avgt    5      7.000              counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE             10         JDK  avgt    5     10.000                  ms
StacktraceRenderBenchmark.render                               SIMPLE             10      CUSTOM  avgt    5     59.304 ±   104.990   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE             10      CUSTOM  avgt    5   3852.650 ±  4215.920  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE             10      CUSTOM  avgt    5    216.003 ±     0.006    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE             10      CUSTOM  avgt    5      5.000              counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE             10      CUSTOM  avgt    5      6.000                  ms
StacktraceRenderBenchmark.render                               SIMPLE           1000         JDK  avgt    5   4370.070 ±  9318.370   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE           1000         JDK  avgt    5   5018.851 ±  6152.230  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE           1000         JDK  avgt    5  19823.986 ±     2.527    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE           1000         JDK  avgt    5      6.000              counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE           1000         JDK  avgt    5      9.000                  ms
StacktraceRenderBenchmark.render                               SIMPLE           1000      CUSTOM  avgt    5   1562.864 ±  2430.437   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE           1000      CUSTOM  avgt    5   7535.696 ±  8585.117  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE           1000      CUSTOM  avgt    5  11216.093 ±     0.144    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE           1000      CUSTOM  avgt    5      8.000              counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE           1000      CUSTOM  avgt    5     12.000                  ms
StacktraceRenderBenchmark.render                               SIMPLE         100000         JDK  avgt    5   3558.101 ±  7396.932   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE         100000         JDK  avgt    5   5806.988 ±  7033.718  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE         100000         JDK  avgt    5  18784.119 ±     0.976    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE         100000         JDK  avgt    5      6.000              counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE         100000         JDK  avgt    5     10.000                  ms
StacktraceRenderBenchmark.render                               SIMPLE         100000      CUSTOM  avgt    5   3526.211 ±  8342.099   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE         100000      CUSTOM  avgt    5   6100.039 ±  8332.617  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE         100000      CUSTOM  avgt    5  18751.963 ±     2.223    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE         100000      CUSTOM  avgt    5      6.000              counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE         100000      CUSTOM  avgt    5      9.000                  ms
StacktraceRenderBenchmark.render                              COMPLEX             10         JDK  avgt    5   6278.604 ± 11966.434   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX             10         JDK  avgt    5   3946.832 ±  5600.596  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX             10         JDK  avgt    5  22352.163 ±     2.149    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX             10         JDK  avgt    5      5.000              counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX             10         JDK  avgt    5      8.000                  ms
StacktraceRenderBenchmark.render                              COMPLEX             10      CUSTOM  avgt    5     52.525 ±    89.409   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX             10      CUSTOM  avgt    5   4333.948 ±  4656.819  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX             10      CUSTOM  avgt    5    216.003 ±     0.005    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX             10      CUSTOM  avgt    5      5.000              counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX             10      CUSTOM  avgt    5      6.000                  ms
StacktraceRenderBenchmark.render                              COMPLEX           1000         JDK  avgt    5   6498.564 ± 13139.639   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX           1000         JDK  avgt    5   3996.477 ±  5477.634  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX           1000         JDK  avgt    5  23352.019 ±     3.477    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX           1000         JDK  avgt    5      5.000              counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX           1000         JDK  avgt    5      7.000                  ms
StacktraceRenderBenchmark.render                              COMPLEX           1000      CUSTOM  avgt    5   1714.024 ±  3431.353   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX           1000      CUSTOM  avgt    5   7171.913 ±  8941.283  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX           1000      CUSTOM  avgt    5  11216.101 ±     0.202    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX           1000      CUSTOM  avgt    5      8.000              counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX           1000      CUSTOM  avgt    5     12.000                  ms
StacktraceRenderBenchmark.render                              COMPLEX         100000         JDK  avgt    5   6462.176 ± 13066.702   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX         100000         JDK  avgt    5   3912.745 ±  5929.656  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX         100000         JDK  avgt    5  22312.247 ±     1.630    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX         100000         JDK  avgt    5      5.000              counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX         100000         JDK  avgt    5      8.000                  ms
StacktraceRenderBenchmark.render                              COMPLEX         100000      CUSTOM  avgt    5   4960.521 ±  4124.533   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX         100000      CUSTOM  avgt    5   4523.745 ±  3719.218  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX         100000      CUSTOM  avgt    5  22799.970 ±     2.906    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX         100000      CUSTOM  avgt    5      5.000              counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX         100000      CUSTOM  avgt    5      7.000                  ms
```

### After
```
StacktraceRenderBenchmark.render                               SIMPLE             10         JDK  avgt    5   3046.303 ±  3689.675   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE             10         JDK  avgt    5   6266.365 ±  6068.397  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE             10         JDK  avgt    5  18824.055 ±     1.155    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE             10         JDK  avgt    5      7.000              counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE             10         JDK  avgt    5     10.000                  ms
StacktraceRenderBenchmark.render                               SIMPLE             10      CUSTOM  avgt    5     66.756 ±   186.035   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE             10      CUSTOM  avgt    5   4778.029 ±  6758.306  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE             10      CUSTOM  avgt    5    264.004 ±     0.011    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE             10      CUSTOM  avgt    5      6.000              counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE             10      CUSTOM  avgt    5     10.000                  ms
StacktraceRenderBenchmark.render                               SIMPLE           1000         JDK  avgt    5   3348.822 ±  4948.949   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE           1000         JDK  avgt    5   6124.539 ±  6295.417  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE           1000         JDK  avgt    5  19823.980 ±     1.982    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE           1000         JDK  avgt    5      7.000              counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE           1000         JDK  avgt    5      9.000                  ms
StacktraceRenderBenchmark.render                               SIMPLE           1000      CUSTOM  avgt    5   1197.760 ±  1532.787   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE           1000      CUSTOM  avgt    5   7241.183 ±  8376.004  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE           1000      CUSTOM  avgt    5   8400.071 ±     0.090    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE           1000      CUSTOM  avgt    5      7.000              counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE           1000      CUSTOM  avgt    5     11.000                  ms
StacktraceRenderBenchmark.render                               SIMPLE         100000         JDK  avgt    5   3629.920 ±  8125.065   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE         100000         JDK  avgt    5   5825.909 ±  7555.525  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE         100000         JDK  avgt    5  18784.215 ±     0.477    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE         100000         JDK  avgt    5      6.000              counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE         100000         JDK  avgt    5      9.000                  ms
StacktraceRenderBenchmark.render                               SIMPLE         100000      CUSTOM  avgt    5   2610.232 ±  5869.223   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                 SIMPLE         100000      CUSTOM  avgt    5   6206.320 ±  7964.692  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm            SIMPLE         100000      CUSTOM  avgt    5  14416.154 ±     0.347    B/op
StacktraceRenderBenchmark.render:gc.count                      SIMPLE         100000      CUSTOM  avgt    5      7.000              counts
StacktraceRenderBenchmark.render:gc.time                       SIMPLE         100000      CUSTOM  avgt    5     10.000                  ms
StacktraceRenderBenchmark.render                              COMPLEX             10         JDK  avgt    5   5012.847 ±  4277.812   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX             10         JDK  avgt    5   4402.214 ±  3185.162  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX             10         JDK  avgt    5  22424.085 ±     1.931    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX             10         JDK  avgt    5      6.000              counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX             10         JDK  avgt    5      8.000                  ms
StacktraceRenderBenchmark.render                              COMPLEX             10      CUSTOM  avgt    5     53.938 ±    44.720   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX             10      CUSTOM  avgt    5   4819.203 ±  3706.502  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX             10      CUSTOM  avgt    5    264.003 ±     0.003    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX             10      CUSTOM  avgt    5      6.000              counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX             10      CUSTOM  avgt    5      8.000                  ms
StacktraceRenderBenchmark.render                              COMPLEX           1000         JDK  avgt    5   6919.014 ± 14549.656   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX           1000         JDK  avgt    5   3934.621 ±  6502.114  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX           1000         JDK  avgt    5  23352.410 ±     0.857    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX           1000         JDK  avgt    5      5.000              counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX           1000         JDK  avgt    5      7.000                  ms
StacktraceRenderBenchmark.render                              COMPLEX           1000      CUSTOM  avgt    5   1395.007 ±  2594.826   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX           1000      CUSTOM  avgt    5   6784.124 ± 10646.415  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX           1000      CUSTOM  avgt    5   8400.082 ±     0.153    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX           1000      CUSTOM  avgt    5      7.000              counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX           1000      CUSTOM  avgt    5     10.000                  ms
StacktraceRenderBenchmark.render                              COMPLEX         100000         JDK  avgt    5   6534.255 ± 14390.630   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX         100000         JDK  avgt    5   3965.385 ±  6360.232  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX         100000         JDK  avgt    5  22216.254 ±     1.661    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX         100000         JDK  avgt    5      5.000              counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX         100000         JDK  avgt    5      8.000                  ms
StacktraceRenderBenchmark.render                              COMPLEX         100000      CUSTOM  avgt    5   3944.260 ±  4426.148   ns/op
StacktraceRenderBenchmark.render:gc.alloc.rate                COMPLEX         100000      CUSTOM  avgt    5   4308.129 ±  4677.541  MB/sec
StacktraceRenderBenchmark.render:gc.alloc.rate.norm           COMPLEX         100000      CUSTOM  avgt    5  16711.803 ±     3.539    B/op
StacktraceRenderBenchmark.render:gc.count                     COMPLEX         100000      CUSTOM  avgt    5      5.000              counts
StacktraceRenderBenchmark.render:gc.time                      COMPLEX         100000      CUSTOM  avgt    5      8.000                  ms
```

File: sdk/common/src/jmh/java/io/opentelemetry/sdk/internal/StacktraceRenderBenchmark.java (hunk 3 above)
Comment (jack-berg) [reply sdk/common/src/jmh/java/io/opentelemetry/sdk/internal/StacktraceRenderBenchmark.java:62]: I'm going for an apples to apples comparison with the current implementation, which is defined [here](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/common/src/main/java/io/opentelemetry/sdk/internal/AttributeUtil.java#L121-L128). Currently, we initialize a StringWriter / PrintWriter each and every time we add an exception to a span! 😬

File: sdk/common/src/main/java/io/opentelemetry/sdk/internal/StackTraceRenderer.java (hunk 4 above)
Comment (jack-berg) [reply sdk/common/src/main/java/io/opentelemetry/sdk/internal/StackTraceRenderer.java:49]: 👍 will do

File: sdk/common/src/main/java/io/opentelemetry/sdk/internal/DefaultExceptionAttributeResolver.java [hunk 5]
```diff
@@ -15,10 +16,19 @@
  */
 public final class DefaultExceptionAttributeResolver implements ExceptionAttributeResolver {
 
+  private static final String ENABLE_JVM_STACKTRACE_PROPERTY =
+      "otel.experimental.sdk.jvm_stacktrace";
+
   private static final DefaultExceptionAttributeResolver INSTANCE =
-      new DefaultExceptionAttributeResolver();
+      new DefaultExceptionAttributeResolver(
+          Boolean.parseBoolean(ConfigUtil.getString(ENABLE_JVM_STACKTRACE_PROPERTY, "false")));
```
Comment (jack-berg) [inline sdk/common/src/main/java/io/opentelemetry/sdk/internal/DefaultExceptionAttributeResolver.java]: Added new config to allow toggling between the naive JVM rendering logic and the new limits aware stacktrace rendering logic. 

Given the solid testing in `StackTraceRendererTest`, I think we should default to the new limits aware rendering logic.

**Comment by jack-berg** [comment]
@open-telemetry/java-approvers please take a look - I'd like to get this in for the 7/11/25 release.

**Review by jkwatson — APPROVED** [review]
Possible to add a little coverage on the uncovered bits? Otherwise, :shipit:

**Comment by jack-berg** [comment]
I think the lack of coverage is a bit of a false positive. The key to hitting those uncovered bits to truncate exception stacktraces in all sorts of odd places. This [test](https://github.com/open-telemetry/opentelemetry-java/pull/7281/files#diff-cfc6903daa0cf66a1b962bce42014f525bb362a8f3382c56fbdd7cf99291cd24R24-R31) does that by introducing a random length limit as a parameter of the test. I think the coverage just didn't happen to randomly choose a limit which exercises those bits, but with enough iterations it should.
